### PR TITLE
WA-497 Added fee estimation update on entering amount or address

### PR DESCRIPTION
**Pull request by DmitryBespalov**

### CODE REVIEW

File: MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift [hunk 1]
```diff
@@ -509,6 +509,22 @@ public class WalletApplicationService: Assertable {
...
+        guard let recipient = formattedAddress, !recipient.isZero else { return nil }
+        let request = EstimateTransactionRequest(safe: Address(selectedWalletAddress!),
+                                                 to: recipient,
+                                                 value: String(amount),
+                                                 data: nil,
+                                                 operation: .call)
+        guard let response = try? DomainRegistry.transactionRelayService.estimateTransaction(request: request) else {
+            return nil
+        }
+        return (BigInt(response.dataGas) + BigInt(response.safeTxGas)) * BigInt(response.gasPrice)
```
Comment (sche) [inline MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift:525]: I think response.dataGas should always be 0 for "data: nil"

File: SafeUIKit/SafeAppUI/AppFlow/MainFlow/FundsTransferTransactionViewController.swift [hunk 2]
```diff
@@ -16,13 +17,90 @@ class FundsTransferTransactionViewController: UIViewController {
...
+
+    fileprivate func estimate(_ amount: String?, _ recipient: String?) -> BigInt? {
+        guard let amountText = amount, let amount = tokenFormatter.number(from: amountText) else { return nil }
+        return ApplicationServiceRegistry.walletService.estimateTransferFee(amount: amount, address: recipient)
+    }
+
+    func updateFeeEstimation(amount: String?, recipient: String?) {
+        userInputQueue.cancelAllOperations()
+        userInputQueue.addOperation(CancellableBlockOperation { [weak self] op in
+            guard let `self` = self else { return }
```
Comment (sche) [inline SafeUIKit/SafeAppUI/AppFlow/MainFlow/FundsTransferTransactionViewController.swift:53]: Two proposals:
1) I think it should be enough just to call one estimation. I think it should not change depending on eth amount.
2) Even if it changes I think it worth to have a delay before first ``` if op.isCancelled { return } ```

File: MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift (hunk 1 above)
Comment (DmitryBespalov) [reply MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift:525]: whatever server returns

File: SafeUIKit/SafeAppUI/AppFlow/MainFlow/FundsTransferTransactionViewController.swift (hunk 2 above)
Comment (DmitryBespalov) [reply SafeUIKit/SafeAppUI/AppFlow/MainFlow/FundsTransferTransactionViewController.swift:53]: yes, I also thought about it.

1) I think it's fine to re-estimate for amount and address, no harm
2) I tried with delay and then researched what other companies doing, and not delaying at all is a common practice (google's and others autocomplete behavior).

Comment (sche) [reply SafeUIKit/SafeAppUI/AppFlow/MainFlow/FundsTransferTransactionViewController.swift:53]: What about boundary cases when some intermediate responses returned some valid estimate, but the last one failed?

Comment (sche) [reply SafeUIKit/SafeAppUI/AppFlow/MainFlow/FundsTransferTransactionViewController.swift:53]: And here the use case is a bit different from what is required in autocomplete

**Pull request merged**